util/src/matrix.rs

```diff
@@ -0,0 +1,527 @@
+            }
+        };
+
+        unsafe {
+            // SAFETY: The closure initializes every element.
+            Self::new_with(width, height, initializer_fn)
+        }
+    }
+
+    /// Transmute each element of the matrix from `T` to `U`.
+    ///
+    /// # Safety
+    ///
+    /// Due to interactions with the allocator, this method has stricter requirements than the
+    /// `mem::transmute` function in the library.
+    ///
+    /// Same as `mem::transmute`:
+    /// - `T` and `U` must have the same size.
+    /// - All elements of the matrix must be safe to interpret bitwise as `U`.
+    ///
+    /// Furthermore:
+    /// - `T` and `U` must have the same alignment.
+    unsafe fn transmute<U>(self) -> Matrix<U> {
+        let Matrix {
+            data,
+            height,
+            width,
+        } = self;
+        let size = data.len();
+
+        let ptr: *mut [T] = Box::into_raw(data);
+        // Warning: we cannot panic from now until we've constructed another box. The buffer is
+        // currently not owned by any object, so it would leak if we have to unwind the stack.
+
+        // Rust inexplicably can't cast to a fat pointer (??) so we cast to a thin pointer to `U`,
+        // then we construct a fat pointer to `U`. (lol)
+        let ptr: *mut U = ptr.cast::<U>();
+        let ptr: *mut [U] = ptr::slice_from_raw_parts_mut(ptr, size);
+
+        // SAFETY: The caller guarantees that the preconditions hold.
+        let data: Box<[U]> = Box::from_raw(ptr);
+
+        Matrix {
+            data,
+            height,
+            width,
+        }
+    }
+
+    /// Wrap elements of the matrix in `ManuallyDrop`, preventing their drop code being run when
+    /// the matrix is dropped.
+    ///
+    /// Although this function is safe, it can cause memory leaks. To avoid them, ensure that all
+    /// resources are correctly dropped in some other way.
+    fn manually_drop_elements(self) -> Matrix<ManuallyDrop<T>> {
+        // SAFETY: `ManuallyDrop<T>` has the same layout as `T`. `T` can be safely reinterpreted as
+        // `ManuallyDrop<T>`.
+        unsafe { self.transmute::<ManuallyDrop<T>>() }
+    }
+}
+
+impl<T> Matrix<MaybeUninit<T>> {
+    /// Convert `Matrix<MaybeUninit<T>>` to `Matrix<T>`, under the assumption that all elements have
+    /// been correcly initialized.
+    ///
+    /// # Safety
+    ///
+    /// All elements of the matrix must have been correctly initialized, such that they can be
+    /// safely transmuted to `T`.
+    pub unsafe fn assume_init(self) -> Matrix<T> {
+        // SAFETY: `MaybeUninit<T>` has the same layout as `T`. `MaybeUninit<T>` can be safely
+        // reinterpreted as `T` when initialized. The caller guarantees that initialization took
+        // place.
+        self.transmute::<T>()
+    }
+}
+
+impl<T: Clone> Clone for Matrix<T> {
+    fn clone(&self) -> Self {
+        Self::from_flat_vec(self.height, self.width, self.as_flat().to_vec())
+    }
+}
```

Comment:
Why not `#[derive(Clone)]`?  `Box<T>` impls clone when T does, so afaik derive should do the right thing.

Reply:
Oh true! That is nicer